text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,517 @@
+    fn by_custom_ptr(self: CustomPtr<Self>);
+}
+
+struct CustomPtr<T>(*const T);
+
+impl<T> Receiver for CustomPtr<T> {
+    type Target = T;
+}
+```
+
+## Recursive arbitrary receivers
+
+Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward). This is the behavior for the current special-cased self types (`Pin`, `Box` etc.) so as we remove the special-casing we need to retain this property.
+
+For example, this self type is valid:
+
+```rust
+impl MyType {
+     fn by_rc_to_box(self: Rc<Box<Self>>) { ... }
+}
+```
+
+The Rust language doesn't provide a way for user code to explore this recursion, so this trait is unlikely to be useful except to the compiler. Nevertheless, we don't intend to _prevent_ use of the `Receiver` trait by user code: since the same recursive property applies to `Deref` yet it's been occasionally useful to [introduce `Deref` bounds](https://doc.rust-lang.org/std/pin/struct.Pin.html#method.new_unchecked).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `core` libs changes
+
+The `Receiver` trait is made public (removing its `#[doc(hidden)])` attribute), exposing it under `core::ops`. It gains a `Target` associated type.
+
+This trait marks types that can be used as receivers other than the `Self` type of an impl or trait definition.
+
+```rust
+pub trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+A blanket implementation is provided for any type that implements `Deref`:
+
+```rust
+impl<P: ?Sized, T: ?Sized> Receiver for P
+where
+    P: Deref<Target = T>,
+{
+    type Target = T;
+}
+```
+
+(See [alternatives](#no-blanket-implementation) for discussion of the tradeoffs here.)
+
+It is also implemented for `&T`, `&mut T`, `*const T` and `*mut T`.
+
+## Compiler changes
+
+The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html). This algorithm changes in one simple way: instead of dereferencing types (using the `Deref<Target=T>`) trait, we use the new `Receiver<Target=T>` trait to determine the next step.
+
+(Note that the existing algorithm isn't quite as simple as following the chain of `Deref<Target=T>`. In particular, `&T` and `&mut T` are considered as candidates too at each step; this RFC does not change that.)
+
+Because a blanket implementation is provided for users of the `Deref` trait and for `&T`/`&mut T`, the net behavior is similar. But this provides the opportunity for types which can't implement `Deref` to act as method receivers.
+
+Dereferencing a raw pointer usually needs `unsafe` (for good reason!) but in this case, no actual dereferencing occurs. This is used only to determine a list of method candidates; no memory access is performed and thus no `unsafe` is needed.
```

Comment:
> > Another note: I'd like to see some more details of how method resolution and deref coercion would work with this change. Currently, it says that the algorithm would remain the same, but instead of using the Deref trait, it would use the new Receiver trait. I'm not sure this is quite right, because some method calls require deref coercions, and this new trait does not imply Deref.
> > 
> > In the following example, the `Deref` trait would still need to be involved to check if the calls to `frobnicate_ref` are allowed.
> > 
> > ```rust
> > struct Foo;
> > 
> > impl Foo{
> >     fn frobnicate_ref(&self) {}
> >     fn frobnicate_cpp_ref(self: CppRef<Self>) {}
> > }
> > 
> > fn main() {
> >     let foo_rc: Rc<Foo> = Rc::new(Foo);
> > 
> >     // this compiles fine, and desugars to `Foo::frobnicate_ref(&*foo_rc)`
> >     foo_rc.frobnicate_ref();
> > 
> >     let foo_cpp_ref: CppRef<Foo> = CppRef::new(Foo);
> > 
> >     // should not compile because it would desugar to `Foo::frobnicate_ref(&*foo_cpp_ref)`
> >     // and you can't deref a CppRef
> >     foo_cpp_ref.frobnicate_ref();
> > 
> >     // should compile, because we're not dereffing the CppRef
> >     // desugars to `Foo::frobnicate_cpp_ref(foo_cpp_ref)`
> >     foo_cpp_ref.frobnicate_cpp_ref();
> > }
> > ```
> 
> Yeah, I caught that too.
> 
> I think the RFC is mixing two sets of semantics at times. As far as I understand, [the reference](https://doc.rust-lang.org/reference/expressions/method-call-expr.html) as it's currently written doesn't explicitly address Box/Arc/etc as receivers, but it does implicitly address them in this paragraph:
> 
> > Then, for each candidate type T, search for a [visible](https://doc.rust-lang.org/reference/visibility-and-privacy.html) method with a receiver of that type in the following places:
> > 
> > * T's inherent methods (methods implemented directly on T).
> > * Any of the methods provided by a [visible](https://doc.rust-lang.org/reference/visibility-and-privacy.html) trait implemented by T. If T is a type parameter, methods provided by trait bounds on T are looked up first. Then all remaining methods in scope are looked up.
> 
> That first bullet point is the part that must be amended to clarify that "T's inherent methods" need to take the Receiver trait into account; the paragraph above about dereferencing chains must be kept untouched.

Huh, yeah, took me like two hours to properly consider this, but I think you've really got the gist of it here! Essentially, `Receiver` should _restrict_ which inherent implementations one can make, but doesn't actually by itself affect method resolution!

How about we change this section to instead be something like:

````suggestion
For the uninitiated, the following impl briefly describes the existing terminology:

```rust
impl ImplementingType {
    fn foo(self: ReceiverType);
}
```

---

Instead of being hard-coded in the compiler, the `Receiver` trait now limits which types are allowed as method receivers. That is, [the allowed types of the `self` parameter in methods](https://doc.rust-lang.org/reference/items/associated-items.html#methods) are now limited to those generated by the following grammar:

```text
P = any type that implements Receiver<Target = S>
S = Self | P
```

The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html). This algorithm changes slightly: instead of only searching `T`'s own inherent methods, any method that has a receiver with the given type is considered.

Note that what the implementing type for a method is (whether that's `T`, `&T`, `CppRef<T>` or `&CppRef<T>`) doesn't affect method resolution, only the receiver type matters in that regard.
````

(The rest of the RFC will likely need changes to accommodate for this, too).

Reply:
Extending the given frobnicate example with more comments, assuming the above change:
```rust
// `foo_rc: Rc<Foo>`, so the possible method candidates are:
// - `Rc<Foo>`
// - `&Rc<Foo>`
// - `&mut Rc<Foo>`
// - `Foo` (by dereferencing)
// - `&Foo`
// - `&mut Foo`
//
// There exists a visible method named `frobnicate_ref` with receiver type `&Foo`, so that is selected.
foo_rc.frobnicate_ref();

// `foo_cpp_ref: CppRef<Foo>`, so the possible method candidates are:
// - `CppRef<Foo>`
// - `&CppRef<Foo>`
// - `&mut CppRef<Foo>`
//
// There are no methods named `frobnicate_ref` with any of the above receiver types, so this is an error.
foo_cpp_ref.frobnicate_ref();

// `foo_cpp_ref` is the same as above, so the method candidates are the same.
//
// In this case though, there exists a visible method named `frobnicate_cpp_ref` with receiver type `&CppRef<Foo>`, so that is selected.
// Note that whether the method is implemented on `Foo`, `CppRef<Foo>` or `&CppRef<Foo>` doesn't matter.
foo_cpp_ref.frobnicate_cpp_ref();
```

This also explains what happens if we were to add the impl below: There would suddenly be two inherent methods with the same name that has receiver type `&CppRef<Foo>`, and that's an error.

```rust
impl<T> CppRef<T> {
    fn frobnicate_cpp_ref(&self) {}
}
```

Reply:
I'm glad to see a more precise Reference section in this RFC, I've felt uncomfortable about it for a while, and I'd like to commit the above suggested change.

The code which is changing for this RFC is indeed in the section of code which assembles method call _candidates_ which I think is another way of saying:
> Essentially, Receiver should restrict which inherent implementations one can make, but doesn't actually by itself affect method resolution!
so I think the understanding here is correct.

@madsmtm please would you have a look at `wfcheck.rs` in the [hypothetical semi-correct implementation of this RFC](https://github.com/rust-lang/rust/compare/master...adetaylor:rust:receiver_trait_with_target_20231117) to see if it matches your renewed understanding? If so I'll go ahead and commit this change, and we can work on the rest of the wording over time.

Reply:
I'm... not really well-enough versed in `rustc` internals to say for sure, though I think it looks correct! (my mental reasoning: `wfcheck.rs` is (amongst a lot of other stuff) checking whether the implementation is allowed or not based on `Receiver`, allowing e.g. `self: CppRef<Self>` but disallowing e.g. `self: u8`).

> The code which is changing for this RFC is indeed in the section of code which assembles method call _candidates_ which I think is another way of saying *snip*

Yup, the receiver candidate selection stays exactly the same, only the inherent method candidate selection changes.

Reply:
OK I had a play around. Our current working branch is somewhat naïve in this respect and we definitely need to be a bit more sophisticated here. I haven't got it right yet. [Here's a comment explaining the remaining stuff that I know we need to figure out.](https://github.com/rust-lang/rust/commit/32c1585788d51d2e6a4fa7e1c0c30cf4461c6477)

I haven't yet convinced myself whether your revised description above is accurate, so I'm not going to commit it yet. I feel that it definitely needs to describe the search algorithm a bit more programmatically. Right now it says "any method that has a receiver of the given type is considered" - but which `impl` blocks does it search to find them?

It's definitely the case that the statement "we just change the algorithm in one simple way" alas isn't quite true though, so we'll definitely make a change here.

Reply:
> but which `impl` blocks does it search to find them?

I think I was kinda thinking... All of them? Though clearly that's ill-defined for cross-crate compilation, so we'll need something more precise here - I am unsure what that should be though, will give it some more thought myself!

Reply:
I also missed a nuance above: In [the original text for allowed `self` types](https://doc.rust-lang.org/reference/items/associated-items.html#methods), `Pin<P>` is actually only allowed for pointer types, e.g. `self: Pin<Self>` is disallowed; unsure if that's something we'd want to preserve?

I'm still fairly convinced that modifying the candidate type search is not the way to go, effectively that part is just saying "which types can I feasibly reach by inserting derefs or a borrow", which is important for when we get to actually desugaring the method call.

Reply:
(Update: this is the bit I'm trying to work on in code right now, so I can validate if the above idea actually works or is nonsense. I'm travelling quite a bit for work at the moment so not getting a lot of time, but hopefully sometime in the next few days we'll find out)